src/metis/http/gist.lua

```diff
@@ -0,0 +1,285 @@
+        write("You need to add a Personal Access Token (PAK) to upload Gists. Follow the instructions at ")
+        setTextColor(colors.blue)
+        write("https://tinyurl.com/GitHubPAK")
+        setTextColor(colors.yellow)
+        write(" to generate one. Make sure to check the '")
+        setTextColor(colors.blue)
+        write("gist")
+        setTextColor(colors.yellow)
+        print("' checkbox on step 7 (under 'Select scopes'). Once done, paste it here.")
+        setTextColor(colors.lime)
+        write("PAK: ")
+        setTextColor(colors.white)
+        local pak = read()
+        if pak == nil or pak == "" then error("Invalid PAK, please try again.") end
+        settings.set("gist.id", pak)
+        settings.save(".settings")
+        headers.Authorization = "token " .. pak
+        return true
+    end
+    return false
+end
+
+-- User API - this can be loaded with require "cc.http.gist"
+
+-- ID can be either just the gist ID or a gist ID followed by a slash and a file name
+-- * If a file name is specified, retrieves that file
+-- * Otherwise, if there's only one file, retrieves that file
+-- * Otherwise, if there's a file named 'init.lua', retrieves 'init.lua'
+-- * Otherwise, if there's more than one file but only one *.lua file, retrieves the Lua file
+-- * Otherwise, retrieves the first Lua file alphabetically (with a warning)
+-- * Otherwise, fails
+
+--- Retrieves one file from a Gist using the specified ID.
+-- @tparam string id The Gist ID to download from. See above comments for more details.
+-- @tparam[opt] function progress A function to use to report status messages.
+-- @treturn string|nil The contents of the specified Gist file, or nil on error.
+-- @treturn string|nil The name of the file that was chosen to be downloaded, or a message on error.
+function gist.get(id, progress)
+    expect(1, id, "string")
+    expect(2, progress, "function", "nil")
+    progress = progress or emptyfn
+    local file
+    if id:find("/") ~= nil then id, file = id:match("^([0-9A-Fa-f:]+)/(.+)$") end
+    if id == nil or not id:match("^[0-9A-Fa-f][0-9A-Fa-f:]+[0-9A-Fa-f]$") then error("bad argument #1 to 'get' (invalid ID)", 2) end
+    if id:find(":") ~= nil then id = id:gsub(":", "/") end
+    progress("Connecting to api.github.com... ")
+    local handle = http.get("https://api.github.com/gists/" .. id)
+    if handle == nil then
+        progress("Failed.\n")
+        return nil, "Failed to connect"
+    end
+    local meta = textutils.unserializeJSON(handle.readAll())
+    local code = handle.getResponseCode()
+    handle.close()
+    if code ~= 200 then
+        progress("Failed.\n")
+        return nil, "Invalid response code (" .. code .. ")" .. (meta and ": " .. meta.message or "")
+    end
+    if meta == nil or meta.files == nil then
+        progress("Failed.\n")
+        return nil, meta and "GitHub API error: " .. meta.message or "Error parsing JSON"
+    end
+    progress("Success.\n")
```

Comment:
Hrmr, worth discussing whether feedback messages should end with a new line or not.

Reply:
The `progress` function is intended to be able to show the same messages as `pastebin`. It was designed around having `progress` be set to `write`, but can be replaced with something else.